### PR TITLE
Parallelize tasks 

### PR DESCRIPTION
**Pull request by Raneet10**
# Description

This PR attempts to parallelize certain tasks in remote VMs, such as restart, update, cleanup.

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] New test case for remote devnet

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

## Testing

- [ ] I have tested this code manually on local environment
- [x] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai

### Manual tests

Tested in a 3 node setup

### CODE REVIEW

File: src/express/commands/cleanup.js [hunk 1]
```diff
@@ -118,29 +140,26 @@ async function startServices(doc) {
...
-    let borUsers = splitToArray(doc['devnetBorUsers'].toString())
     let user, ip
+    user = `${doc['ethHostUser']}`
+    ip = `${user}@${doc['devnetBorHosts'][0]}`
 
-    for (let i = 0; i < doc['devnetBorHosts'].length; i++) {
-
-        i === 0 ? user = `${doc['ethHostUser']}` : user = `${borUsers[i]}`
-        ip = `${user}@${doc['devnetBorHosts'][i]}`
+    console.log("📍Deploying contracts for bor on machine " + ip + " ...")
```
Comment (marcello33) [inline src/express/commands/cleanup.js:155]: Well done removing the `for` cycle

File: src/express/common/config-utils.js [hunk 2]
```diff
@@ -107,6 +107,10 @@ export function splitToArray(value) {
     return value.split(' ').join('').split(",")
 }
 
+export function splitIp(value) {
```
Comment (marcello33) [inline src/express/common/config-utils.js:110]: As we use this only with the `hostIp`, can we rename the method to `splitArrayAndGetHostIp` and make it return `value.split("@")[0]`?
Also, please introduce an error handling check.
If value passed in is not an array or is undefined, the program should log an error and exit with status code `1`

File: src/express/commands/update.js [hunk 3]
```diff
@@ -145,16 +165,26 @@ export async function updateHeimdall(n) {
...
+
+        let updateHeimdallTasks = nodeIps.map(async(ip) => {
+            await pullAndRestartHeimdall(ip, hostToIndex.get(ip), true)
+        })
+
+        await Promise.all(updateHeimdallTasks)
+
     } else {
-        vmIndex === 0 ? user = `${doc['ethHostUser']}` : `${borUsers[vmIndex]}`
+        vmIndex === 0 ? user = `${doc['ethHostUser']}` : user = `${borUsers[vmIndex]}`
```
Comment (marcello33) [inline src/express/commands/update.js:187]: good catch

File: src/express/commands/update.js [hunk 4]
```diff
@@ -123,16 +133,26 @@ export async function updateBor(n) {
...
+
+        let updateBorTasks = nodeIps.map(async(ip) => {
+            await pullAndRestartBor(ip, hostToIndex.get(ip), true)
+        })
+
+        await Promise.all(updateBorTasks)
+
     } else {
-        vmIndex === 0 ? user = `${doc['ethHostUser']}` : `${borUsers[vmIndex]}`
+        vmIndex === 0 ? user = `${doc['ethHostUser']}` : user = `${borUsers[vmIndex]}`
```
Comment (marcello33) [inline src/express/commands/update.js:155]: good catch

File: src/express/commands/start.js [hunk 5]
```diff
@@ -252,13 +252,13 @@ async function eventuallyCleanupPreviousDevnet(ips, devnetType) {
...
 
-    await Promise.all(cleanup)
+    await Promise.all(cleanupTasks)
 
 }
 
 async function runDockerSetupWithMaticCLI(ips) {
     let doc = await yaml.load(fs.readFileSync('./configs/devnet/remote-setup-config.yaml', 'utf8'));
-    let ipsArray = ips.split(' ').join('').split(",")
+    let ipsArray = splitToArray(ips)
```
Comment (marcello33) [inline src/express/commands/start.js:261]: +1

File: src/express/commands/start.js [hunk 6]
```diff
@@ -314,7 +314,7 @@ async function runDockerSetupWithMaticCLI(ips) {
 async function runRemoteSetupWithMaticCLI(ips) {
 
     let doc = await yaml.load(fs.readFileSync('./configs/devnet/remote-setup-config.yaml', 'utf8'));
-    let ipsArray = ips.split(' ').join('').split(",")
+    let ipsArray = splitToArray(ips)
```
Comment (marcello33) [inline src/express/commands/start.js:317]: +1

File: src/express/commands/start.js [hunk 7]
```diff
@@ -46,8 +46,8 @@ async function installRequiredSoftwareOnRemoteMachines(ips, devnetType) {
     
     let arr = []
 
-    let deps = nodeIps.map(async(ip) => {
-        arr = ip.split("@")
+    let depTasks = nodeIps.map(async(ip) => {
```
Comment (marcello33) [inline src/express/commands/start.js:49]: can we call this variable `requirementTasks` ?

File: src/express/commands/restart.js [hunk 8]
```diff
@@ -12,15 +12,25 @@ export async function restartAll(n) {
 
     let doc = await yaml.load(fs.readFileSync('./configs/devnet/remote-setup-config.yaml', 'utf8'));
     let borUsers = splitToArray(doc['devnetBorUsers'].toString())
+    let nodeIps = []
+    let hostToIndex = new Map()
```
Comment (marcello33) [inline src/express/commands/restart.js:16]: just for consistency, can we name this `hostToIndexMap` ?

**Review by marcello33 — CHANGES_REQUESTED** [review]

File: src/express/common/config-utils.js (hunk 2 above)
Comment (Raneet10) [reply src/express/common/config-utils.js]: Done: [9357450](https://github.com/maticnetwork/matic-cli/pull/86/commits/9357450e975cf3541ff7bdf3e9d0eb979b88726f)

File: src/express/commands/restart.js (hunk 8 above)
Comment (Raneet10) [reply src/express/commands/restart.js]: Fixed: [b658a55](https://github.com/maticnetwork/matic-cli/pull/86/commits/b658a55711f29b38b146399cdecd63cb4ecab326)

File: src/express/commands/start.js (hunk 7 above)
Comment (Raneet10) [reply src/express/commands/start.js]: Fixed: [b658a55](https://github.com/maticnetwork/matic-cli/pull/86/commits/b658a55711f29b38b146399cdecd63cb4ecab326)

**Pull request merged**